imagehash.py

```diff
@@ -372,3 +376,219 @@ def colorhash(image, binbits=3):
+		segmentation_image_size=300
+	):
+	"""
+	Creates a CropResistantHash object, by the algorithm described in the paper "Efficient Cropping-Resistant Robust
+	Image Hashing". DOI 10.1109/ARES.2014.85
+	This algorithm partitions the image into bright and dark segments, using a watershed-like algorithm, and then does
+	an image hash on each segment. This makes the image much more resistant to cropping than other algorithms, with
+	the paper claiming resistance to up to 50% cropping, while most other algorithms stop at about 5% cropping.
+	:param image: The image to hash
+	:param hash_func: The hashing function to use
+	:param limit_segments: If you have storage requirements, you can limit to hashing only the M largest segments
+	:param segment_threshold: Brightness threshold between hills and valleys. This should be static, putting it between
+	peak and trough dynamically breaks the matching
+	:param min_segment_size: Minimum number of pixels for a hashable segment
+	:param segmentation_image_size: Size which the image is resized to before segmentation
+	"""
+	# Define some tunable params
+	gaussian_blur = 2
+	median_filter = 3
+	if hash_func is None:
+		hash_func = dhash
+
+	orig_image = image.copy()
+	# Convert to gray scale and resize
+	image = image.convert("L").resize((segmentation_image_size, segmentation_image_size), Image.ANTIALIAS)
+	# Add filters
+	image = image.filter(ImageFilter.GaussianBlur(gaussian_blur)).filter(ImageFilter.MedianFilter(median_filter))
+	pixels = numpy.array(image).astype(numpy.float32)
+
+	# threshold pixels
+	threshold_pixels = pixels > segment_threshold
+	unassigned_pixels = numpy.full(pixels.shape, True, dtype=numpy.bool)
+
+	segments = []
+	already_segmented = set()
+
+	# Add all the pixels around the border outside the image:
+	already_segmented.update([(-1, z) for z in range(segmentation_image_size)])
+	already_segmented.update([(z, -1) for z in range(segmentation_image_size)])
+	already_segmented.update([(segmentation_image_size, z) for z in range(segmentation_image_size)])
+	already_segmented.update([(z, segmentation_image_size) for z in range(segmentation_image_size)])
+
+	# Find all the "hill" regions
+	while numpy.bitwise_and(threshold_pixels, unassigned_pixels).any():
+		remaining_pixels = numpy.bitwise_and(threshold_pixels, unassigned_pixels)
+		segment = _find_region(remaining_pixels, already_segmented)
+		# Apply segment
+		if len(segment) > min_segment_size:
+			segments.append(segment)
+		for pix in segment:
+			unassigned_pixels[pix] = False
+
+	# Invert the threshold matrix, and find "valleys"
+	threshold_pixels_i = numpy.invert(threshold_pixels)
+	while len(already_segmented) < segmentation_image_size * segmentation_image_size:
+		remaining_pixels = numpy.bitwise_and(threshold_pixels_i, unassigned_pixels)
+		segment = _find_region(remaining_pixels, already_segmented)
+		# Apply segment
+		if len(segment) > min_segment_size:
+			segments.append(segment)
+		for pix in segment:
```

Comment:
Does this need to respect the if just above?

Reply:
No, if the region is too small to be hashed, we still want to consider those pixels to have been checked already, rather than re-checking them